pkg/test/e2e/awaitility.go

```diff
@@ -0,0 +1,136 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/kubefed/pkg/apis/core/common"
+	"sigs.k8s.io/kubefed/pkg/apis/core/v1beta1"
+	"sigs.k8s.io/kubefed/pkg/controller/util"
+	"testing"
+	"time"
+)
+
+// Awaitility contains information necessary for verifying availability of resources in both operators
+type Awaitility struct {
+	T        *testing.T
+	Client   framework.FrameworkClient
+	MemberNs string
+	HostNs   string
+}
+
+// SingleAwaitility contains information necessary for verifying availability of resources in a single operators
+type SingleAwaitility struct {
+	T      *testing.T
+	Client framework.FrameworkClient
+	Ns     string
+}
+
+// Member creates SingleAwaitility for the member operator
+func (a *Awaitility) Member() *SingleAwaitility {
+	return &SingleAwaitility{
+		T:      a.T,
+		Client: a.Client,
+		Ns:     a.MemberNs,
+	}
+}
+
+// Host creates SingleAwaitility for the host operator
+func (a *Awaitility) Host() *SingleAwaitility {
+	return &SingleAwaitility{
+		T:      a.T,
+		Client: a.Client,
+		Ns:     a.HostNs,
+	}
+}
+
+// ReadyKubeFedCluster is a ClusterCondition that represents cluster that is ready
+var ReadyKubeFedCluster = &v1beta1.ClusterCondition{
+	Type:   common.ClusterReady,
+	Status: v1.ConditionTrue,
+}
+
+// WaitForReadyKubeFedClusters waits until both KubeFedClusters (host and member) exist and has ready ClusterCondition
+func (a *Awaitility) WaitForReadyKubeFedClusters() error {
+	if err := a.Host().WaitForKubeFedCluster(a.MemberNs, cluster.Member, ReadyKubeFedCluster); err != nil {
+		return err
+	}
+	if err := a.Member().WaitForKubeFedCluster(a.HostNs, cluster.Host, ReadyKubeFedCluster); err != nil {
+		return err
+	}
+	return nil
+}
+
+// WaitForKubeFedCluster waits until there is a KubeFedCluster representing a operator of the given type
+// and running in the given expected namespace. If the given condition is not nil, then it also checks
+// if the CR has the ClusterCondition
+func (a *SingleAwaitility) WaitForKubeFedCluster(expNs string, clusterType cluster.Type, condition *v1beta1.ClusterCondition) error {
+	timeout := Timeout
+	if condition != nil {
+		timeout = (util.DefaultClusterHealthCheckPeriod + 5) * time.Second
+	}
+	return wait.Poll(RetryInterval, timeout, func() (done bool, err error) {
+		_, ok, err := a.GetKubeFedCluster(expNs, clusterType, condition)
+		if ok {
+			return true, nil
+		}
+		a.T.Logf("waiting for availability of %s KubeFedCluster CR (in namespace %s) representing operator running in namespace '%s'", clusterType, a.Ns, expNs)
+		return false, err
+	})
+}
+
+// WaitForKubeFedClusterWithName waits until there is a KubeFedCluster with the given name
+// and with the given ClusterCondition (if it the condition is nil, then it skips this check)
+func (a *SingleAwaitility) WaitForKubeFedClusterWithName(name string, condition *v1beta1.ClusterCondition) error {
+	timeout := Timeout
+	if condition != nil {
+		timeout = (util.DefaultClusterHealthCheckPeriod + 5) * time.Second
```

Comment:
the DefaultClusterHealthCheckPeriod is currently 10 seconds and to update the ClusterCondition it usually takes 11-12 seconds. I put the 3 additional seconds just as a cushion to not fail just because it didn't make to change the condition in time. I can lower the cushion to 3 seconds, but that won't help much with making the test faster. What would help is to have the periods customizable - I created a Jira issue for that: https://jira.coreos.com/browse/CRT-216